src/global_scheduler/global_scheduler_algorithm.cc

```diff
@@ -111,23 +112,97 @@ double calculate_cost_pending(const GlobalSchedulerState *state,
   locally_available_data_size(state, scheduler->id, task_spec);
   /* TODO(rkn): This logic does not load balance properly when the different
    * machines have different sizes. Fix this. */
-  return scheduler->num_recent_tasks_sent + scheduler->info.task_queue_length;
+  double cost_pending = scheduler->num_recent_tasks_sent +
+      scheduler->info.task_queue_length - scheduler->info.available_workers;
+  return cost_pending;
 }
 
-bool handle_task_waiting(GlobalSchedulerState *state,
+bool handle_task_waiting_random(GlobalSchedulerState *state,
```

Comment:
For some reason this function is very slow. Replacing it with a trivial version

```c++
bool handle_task_waiting_random(GlobalSchedulerState *state,
                                GlobalSchedulerPolicyState *policy_state,
                                Task *task) {
  assign_task_to_local_scheduler(state, task, task->local_scheduler_id);
  return true;
}
```

speeds the benchmark I'm looking at up a ton (just submit 10000 trivial tasks and get the results).

I don't think it's the random number generation or logging... but not 100% sure what it is.

Reply:
policy correctness was my first priority. I'll look into perf in the morning.